soccer/planning/CompositePath.cpp

```diff
@@ -0,0 +1,115 @@
+#include "CompositePath.hpp"
+
+using namespace std;
+namespace Planning
+{
+	CompositePath::CompositePath(unique_ptr<Path> path) {
+		append(std::move(path));
+	}
+
+	CompositePath::CompositePath(Path *path) {
+		append(path);
+	}
+
+	void CompositePath::append(Path *path) {
+		append(std::unique_ptr<Path>(path));
+	}
+
+	void CompositePath::append(unique_ptr<Path> path) {
+		if (path && path->getDuration()>0){
+			paths.push_back(std::move(path));
+		}
+	}
+
+	bool CompositePath::evaluate(float t, Geometry2d::Point &targetPosOut, Geometry2d::Point &targetVelOut) const 
+	{
+		if (paths.empty()) {
+			return false;
+		}
+		if (t<0) {
+			return false;
+		}
+		for (const std::unique_ptr<Path> &path: paths)
+		{
+			float timeLength = path->getDuration();
+			t -= timeLength;
+			if (t<=0 || timeLength == -1) {
+				t += timeLength;
+				path->evaluate(t, targetPosOut, targetVelOut);
+				return true;
+			}
+		}
+		return false;
+	}
+
+	bool CompositePath::hit(const Geometry2d::CompositeShape &shape, float startTime) const
+	{
+		if (paths.empty()) {
+			return false;
+		}
+		int start = 0;
+		for (const std::unique_ptr<Path> &path: paths)
+		{
+			start++;
+			float timeLength = path->getDuration();
+			if (timeLength == -1) {
+				return path->hit(shape, startTime);
+			}
+			startTime -= timeLength;
+			if (startTime<=0) {
+				startTime += timeLength;
+				if (path->hit(shape, startTime)) {
+					return true;
+				}
+				break;
+			}
+		}
+
+		for (;start<paths.size(); start++) {
+			if (paths[start]->hit(shape)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	void CompositePath::draw(SystemState * const state, const QColor &color, const QString &layer) const
```

Comment:
I mean I guess its not really necessary for the state variable, but is there a reason not to? It just enforces a little extra thing to make sure we don't somehow lose that pointer in our draw function.